include/boost/rational.hpp

```diff
@@ -148,33 +152,102 @@ class rational :
-    rational& assign(param_type n, param_type d);
+    template <class T>
+    typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value, rational &
+    >::type assign(const T& n, const T& d)
+    {
+       return *this = rational<IntType>(static_cast<IntType>(n), static_cast<IntType>(d));
+    }
+    //
+    // The following overloads should probably *not* be provided - 
+    // but are provided for backwards compatibity reasons only.
+    // These allow for construction/assignment from types that
+    // are wider than IntType only if there is an implicit
+    // conversion from T to IntType, they will throw a bad_rational
+    // if the conversion results in loss of precision or undefined behaviour.
+    //
+    template <class T>
+    rational(const T& n, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, static_cast<T>(1));
+    }
+    template <class T>
+    rational(const T& n, const T& d, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, d);
+    }
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type operator=(const T& n) { return assign(n, static_cast<T>(1)); }
+
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type assign(const T& n, const T& d)
+    {
+       if(!is_safe_narrowing_conversion(n) || !is_safe_narrowing_conversion(d))
+          BOOST_THROW_EXCEPTION(bad_rational());
```

Comment:
> I agree that the same reasoning applies. The only two sensible 
> behaviors I see for op+-*/ on overflow are to either throw, or return 
> the closest representable. (For op- with an underlying unsigned type, 
> throwing when the result would be negative looks like the only option.)
>
> This does present a bit of a problem with op++/-- though, because an 
> overflow/underflow check there would be measurable, so it should 
> probably be skipped. The special case of +- with the same denominator 
> would also be slowed down by the check.
>
> I agree that this is not a trivial decision.
>

I think we're mixing two different issues here - the point of this patch 
is close some loopholes whereby narrowing conversions - in particular 
truncating-float-to-rational were previously allowed *by accident*.  In 
closing that one, I had to decide what to do about narrowing integer 
conversions as well.  These were also allowed by accident, not design, 
but I felt that there was too much potential to break old code by 
completely disallowing them.  Hence the runtime checks.

Checking in the operators is a whole other ball game:

* It's non-trivial (in comparison to the constructors), such checks 
would require integer division if implemented portably, and in the 
multiprecision case, that's ruinously expensive.
* It would be darn hard to implement correctly - we would need a checked 
version of lcm at the very least.
* It's outside the scope of this patch, and frankly there are other 
pieces of low hanging fruit that should be fixed first (non-throwing 
move construct/assign would get my vote).
* Now that we have Boost .Multiprecision there are other options - 
unbounded integer types, or bounded and checked multiprecision ones. The 
latter can check for overflow at virtually no cost, certainly much 
cheaper than portable checking code in the operators could ever do.

However, Steven is correct that I missed one other converting 
constructor: I've just added runtime checks to that as well for 
consistency.  I'll also add some compile-fail tests shortly to verify 
that stuff that should no longer compiler really doesn't.